java.hints/src/org/netbeans/modules/java/hints/jdk/ConvertToVarHint.java

```diff
@@ -148,9 +172,17 @@ private static boolean preConditionChecker(HintContext ctx) {
         }
 
         TreePath treePath = ctx.getPath();
+        EnhancedForLoopTree efl = null;
+        TypeMirror expressionType = null;
+        if (treePath.getLeaf().getKind() == Tree.Kind.ENHANCED_FOR_LOOP) {
+            efl = (EnhancedForLoopTree) treePath.getLeaf();
+            expressionType = ctx.getInfo().getTrees().getTypeMirror(new TreePath(treePath, efl.getExpression()));
+        }
 
+        Element treePathElement = info.getTrees().getElement(treePath);
         // variable should have local scope
-        if (info.getTrees().getElement(treePath).getKind() != ElementKind.LOCAL_VARIABLE) {
+        //No need to DeclaredType check as primitive types can also be replaced with var in enhanced-for-loop
+        if ((treePathElement != null && treePathElement.getKind() != ElementKind.LOCAL_VARIABLE) && expressionType == null) {
```

Comment:
If I (this time) understood well, the expressionType `!= null` for enhanced for-loop.  So the left side of condition only applies in variable declaration case. To be more explicit, it could move to `else` branch of the if:
```
 if (treePath.getLeaf().getKind() == Tree.Kind.ENHANCED_FOR_LOOP) {
...
} else {
    Element treePathElement = info.getTrees().getElement(treePath);
    if ((treePathElement != null && treePathElement.getKind() != ElementKind.LOCAL_VARIABLE) {
}

Reply:
additional valid context for the hint is `treePathElement.getKind() == ElementKind.RESOURCE_VARIABLE`